torch/_tensor.py

```diff
@@ -96,7 +96,14 @@ def __deepcopy__(self, memo):
             # doesn't work because of
             # https://github.com/pytorch/pytorch/issues/47442
             # Update the test in test_serialization if you remove 'meta' from here
-            if self.is_sparse or self.device.type in ['lazy', 'xla', 'mps', 'ort', 'meta', 'hpu'] or \
+
+            fake_tensor_dispatch = isinstance(
```

Comment:
Could we try something like:
```
class FakeTensorMode:
  def __init__(...):
    self.dispatch_mode = FakeTensorDispatchMode()  # the current FakeTensorMode
    self.function_mode = FakeTensorFunctionMode()  # one that only defines a torch_function
  def __enter__(...):
     with self.dispatch_mode:
       with self.function_mode:
          yield
```

then use this like `with FakeTensorMode()` (this is a new API that we're trying to switch over to lol sorry for all the changes)

Reply:
When I try just a pretty small self-contained version I get an error:
```
import torch
import copy
from torch.overrides import TorchFunctionMode


class ModeTensor(torch.Tensor):
    def __new__(cls, elem, mode):
        r = torch.Tensor._make_subclass(cls, elem, elem.requires_grad)
        r.elem = elem
        r.mode = mode
        return r

    def __torch_function__(self, func, types, args=(), kwargs=None):
        with self.mode:
            return func(*args, **kwargs)

class Mode(TorchFunctionMode):
    def __torch_function__(self, func, types, args=(), kwargs=None):
        def unwrap(e):
            if isinstance(e, ModeTensor):
                return e.elem
            else:
                return e

        def wrap(t):
            if isinstance(t, torch.Tensor):
                return ModeTensor(t, self)
            else:
                return t

        kwargs = kwargs if kwargs else {}
        return wrap(func(*tuple(unwrap(a) for a in args), **kwargs))

x = torch.rand([4])
mod = torch.nn.Linear(10, 10)
with Mode():
    mod2 = copy.deepcopy(mod)
print(mod2)
print(mod2.weight)
```

> Traceback (most recent call last):
  File "/scratch/eellison/elias.py", line 39, in <module>
    print(mod2.weight)
  File "/private/home/eellison/anaconda3/lib/python3.8/site-packages/torch/_tensor.py", line 350, in __repr__
    return handle_torch_function(Tensor.__repr__, (self,), self,
  File "/private/home/eellison/anaconda3/lib/python3.8/site-packages/torch/overrides.py", line 1522, in handle_torch_function
    result = torch_func_method(public_api, types, args, kwargs)
  File "/scratch/eellison/elias.py", line 14, in __torch_function__
    with self.mode:
  File "/private/home/eellison/anaconda3/lib/python3.8/site-packages/torch/overrides.py", line 1847, in __enter__
    raise RuntimeError(f"{self} has already been used as a mode. Please use a fresh version or use restore")
RuntimeError: <__main__.Mode object at 0x7fc0db542820> has already been used as a mode. Please use a fresh version or use restore

Reply:
(same thing regardless if I use `inner=None` or `with mode.push()`

Reply:
The changes here are extremely minimal (~5 lines) and I would rather not block on fixing internal TorchDispatch errors or figuring the inheritance of both TorchFunctionMode and TorchDispatchMode which doesnt exist today, especially since the APIs are still in flux.

Reply:
Ahh--could you try that but `with self.mode.restore()`

Reply:
> The changes here are extremely minimal (~5 lines) and I would rather not block on fixing internal TorchDispatch errors or figuring the inheritance of both TorchFunctionMode and TorchDispatchMode which doesnt exist today, especially since the APIs are still in flux.

I think the idea of leaking details about the existence of mode, let alone the specific FakeTensor mode, to tensor internals is pretty bad practice* and is directly backwards of how we generally want the relationship of modes and functions (modes should know about functions, not the other way around). If there's any way that we can make this work without changing the tensor implementation itself, it seems preferable

* Specifically if someone had a limited build that excludes fake_tensor.py, they can no longer use pytorch at all

Reply:
```
Traceback (most recent call last):
  File "/scratch/eellison/elias.py", line 40, in <module>
    print(mod2.weight)
  File "/private/home/eellison/anaconda3/lib/python3.8/site-packages/torch/_tensor.py", line 350, in __repr__
    return handle_torch_function(Tensor.__repr__, (self,), self,
  File "/private/home/eellison/anaconda3/lib/python3.8/site-packages/torch/overrides.py", line 1522, in handle_torch_function
    result = torch_func_method(public_api, types, args, kwargs)
  File "/scratch/eellison/elias.py", line 14, in __torch_function__
    with self.mode.restore():
  File "/private/home/eellison/anaconda3/lib/python3.8/contextlib.py", line 113, in __enter__
    return next(self.gen)
  File "/private/home/eellison/anaconda3/lib/python3.8/site-packages/torch/utils/_mode_utils.py", line 96, in _restore_mode
    raise RuntimeError(f"{mode} does not have any ancestors. Use the standard version instead of restore")
RuntimeError: <__main__.Mode object at 0x7f5359fc5f40> does not have any ancestors. Use the standard version instead of restore
```

Reply:
I will try to introduce a separate Mode class for deepcopy, although I think the api is a bit wonky.

Reply:
I agree it's wonky and generally the concept of having some interception at the torch function and other interception at the torch dispatch level isn't something we'd considered. Let me know if you need help writing it. Also JFYI that I wasn't able to repro the second error and I'm not certain what's up with that

Reply:
So, these are the functions that get called if you call deepcopy are:
```
<method-wrapper '__get__' of getset_descriptor object at 0x7efb77191f80>
<method 'clone' of 'torch._C._TensorBase' objects>
<function Tensor.__hash__ at 0x7efb76fda3a0>
<method-wrapper '__get__' of getset_descriptor object at 0x7efb77193280>
<method 'detach' of 'torch._C._TensorBase' objects>
<method 'requires_grad_' of 'torch._C._TensorBase' objects>
<method-wrapper '__get__' of getset_descriptor object at 0x7efb77191f80>
<method 'clone' of 'torch._C._TensorBase' objects>
<function Tensor.__hash__ at 0x7efb76fda3a0>
<method-wrapper '__get__' of getset_descriptor object at 0x7efb77193280>
<method 'detach' of 'torch._C._TensorBase' objects>
<method 'requires_grad_' of 'torch._C._TensorBase' objects>
<method-wrapper '__get__' of getset_descriptor object at 0x7efb77191f80>
<method 'clone' of 'torch._C._TensorBase' objects>
<function Tensor.__hash__ at 0x7efb76fda3a0>
<method-wrapper '__get__' of getset_descriptor object at 0x7efb77193280>
<method 'detach' of 'torch._C._TensorBase' objects>
<method 'requires_grad_' of 'torch._C._TensorBase' objects>
```

The following seems to work:
```
class Mode(TorchFunctionMode):
    def __init__(self, fake_mode):
        self.fake_mode = fake_mode

    def __torch_function__(self, func, types, args=(), kwargs=None):
        def wrap(t):
            if isinstance(t, torch.Tensor) and not isinstance(t, FakeTensor):
                return self.fake_mode.from_tensor(t)
            else:
                return t

        kwargs = kwargs if kwargs else {}
        return wrap(func(*args, **kwargs))
```

but I'm not convinced it's 100p correct.. either way going with that for now I guess

Reply:
Hmm this doesnt copy over `.grad`...  😢

Reply:
In the mode, can we just check if the func is `Tensor.__deepcopy__` and then have that just return `self.fake_mode.from_tensor(self)` and not call the function again since that's the behavior that's it looks like was originally here

Reply:
The original tensor is not FakeTensor, so `Tensor.__deepcopy__` does not get called. I think I need to add a check if a `TorchFunctionMode` is active here https://github.com/pytorch/pytorch/blob/master/torch/_tensor.py#L86 and call `__deepcopy__` on the mode with tensors

Reply:
`has_torch_function_unary ` [has a check that the mode is on](https://github.com/pytorch/pytorch/blob/master/torch/csrc/utils/disable_torch_function.cpp#L223)...really weirdly, with some debugging, I'm not even seeing `__deepcopy__` being called in this case (if I add a breakpoint at the top of that function, it never gets hit)

Reply:
Okay so what's happening is that [Parameter has a custom deepcopy](https://github.com/pytorch/pytorch/blob/master/torch/nn/parameter.py#L52) that's being called and it doesn't redispatch to tensor's deepcopy. I'm a little confused how it was working before with a change to `_tensor.py`?

To double check, the main use case for this is in fact modules?

If so, a couple thoughts:
- Do we want to always continue tracking gradients through clone? If so, we should be able to have special behavior for clone in the dispatch mode that doesn't redispatch to `.clone()` and just uses the same underlying tensor
- Otherwise, it might be okay to add a funky `check_torch_function` to parameter's deep copy that does like
```
if _get_torch_function_mode():
  res_data = handle_torch_function('Tensor.__deep_copy__', ...)
else:
  res_data = self.data.clone(memory_format=torch.preserve_format)
result = type(self)(res_data, self.requires_grad)
```

However, if we go with the second option, I would really want to know why we want to use the same tensor in the deep copy case and not in other cases.

cc @zou3519 for any other brainstorming ideas/vetos of current ideas

Reply:
In my current diff: I added
```
if _get_torch_function_mode():
      return handle_torch_function(Parameter.__deepcopy__, (self,), self, memo)
```` 

which is pretty similar

Reply:
If we do this, note that we will now be returning a FakeTensor(Parameter( instead of a Parameter(FakeTensor()) and it looks like the tests [are expecting different behavior](https://github.com/pytorch/pytorch/pull/79580/files#diff-3bc0cf540b694f4ec0a3749f78b047456657a53a5657e495ffb68e5970c5fdaaR167). Also I'm still unclear why we want different behavior for deep copy than with normal clone

Reply:
> note that we will now be returning a FakeTensor(Parameter( instead of a Parameter(FakeTensor())

I think that's dependent on the implementation - that's now it's working rn afaik.